docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/private-api-is-used-on-the-project-level.md

```diff
@@ -128,3 +130,416 @@ class CustomerReader extends SprykerCustomerReader implements CustomerReaderInte
+
+`CustomerAccessFilter` uses as dependency `CustomerAccessReaderInterface` that was declared on the vendor level.
+
+```php
+...
+class CustomerAccessFilter implements CustomCustomerAccessFilter
+{
+    /**
+    * @param \Pyz\Zed\CustomerAccess\CustomerAccessConfig $customerAccessConfig
+    * @param \Spryker\Zed\CustomerAccess\Business\CustomerAccess\CustomerAccessReaderInterface $customerAccessReader
+    */
+    public function __construct(CustomerAccessConfig $customerAccessConfig, CustomerAccessReaderInterface $customerAccessReader)
+    {
+        $this->customerAccessReader = $customerAccessReader;
+        $this->customerAccessConfig = $customerAccessConfig;
+    }
+}
+```
+
+### Example of related error in the Evaluator output
+
+```bash
+------------------------------------------------------------------------------------------------------------------------
+PrivateApi:DependencyInBusinessModel "Please avoid usage of Spryker\Zed\CustomerAccess\Business\CustomerAccess\CustomerAccessReaderInterface in Pyz\Zed\CustomerAccess\Business\CustomerAccess\CustomerAccessFilter"
+------------------------------------------------------------------------------------------------------------------------
+```
+
+### Resolving the error
+
+The solution for the error is the same as in `Resolving the error`[PrivateApi:Facade](#privateapi-facade)
+
+
+## PrivateApi:ObjectInitialization
+
+Business models on the project level should avoid creating private API from the Core level.
+
+### Example of code that causes an upgradability error
+
+`CustomerAccessFilter` contains creating `CustomerAccessReader` in the `constructor` method.
+
+```php
+...
+class CustomerAccessFilter implements CustomCustomerAccessFilter
+{
+    public function __construct(
+        CustomerAccessEntityManagerInterface $customerAccessEntityManager,
+        CustomerAccessReaderInterface $customerAccessReader,
+        CustomerAccessFilterInterface $customerAccessFilter
+    ) {
+        parent::__construct($customerAccessEntityManager);
+        $this->customerAccessReader = new CustomerAccessReader();
+        $this->customerAccessFilter = $customerAccessFilter;
+    }
+}
+```
+
+### Example of related error in the Evaluator output
+
+```bash
+------------------------------------------------------------------------------------------------------------------------
+Please avoid usage of Spryker\Zed\CustomerAccess\Business\CustomerAccess\CustomerAccessReaderInterface in Pyz\Zed\CustomerAccess\Business\CustomerAccess\CustomerAccessFilter
+------------------------------------------------------------------------------------------------------------------------
+```
+
+### Resolving the error
+
+The solution for the error is the same as in `Resolving the error`[PrivateApi:Facade](#privateapi-facade)
+
+
+## PrivateApi:PersistenceInBusinessModel
+
+Business models on project level can use repository and entity manager but only methods that were declared on project level.
```

Comment:
How do methods relate to repository and entity managers? My guess: Business models can extend repository and entity manager from the private API, but their methods must be declared on the project level.

Reply:
It is close to PrivateApi:Facade

What about the version?

When you use a repository, factory, or entity manager inside the business model classes on a project level, please use only methods that are also declared on the project level.